File: backend/app/api/v1/tasks.py
from uuid import UUID

from fastapi import APIRouter, Depends, Query, Response, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.dependencies import get_current_user
from app.db.session import get_db
from app.models.user import User
from app.schemas.task import TaskCreate, TaskResponse, TaskUpdate
from app.services.task_service import (
    archive_task,
    create_task,
    get_task_or_404,
    list_tasks,
    update_task,
)

router = APIRouter()


@router.post("/", response_model=TaskResponse, status_code=status.HTTP_201_CREATED)
async def create_task_route(
    payload: TaskCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> TaskResponse:
    task = await create_task(db, current_user, payload)
    return TaskResponse.model_validate(task)


@router.get("/", response_model=list[TaskResponse])
async def list_tasks_route(
    include_archived: bool = Query(default=False),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> list[TaskResponse]:
    tasks = await list_tasks(db, current_user, include_archived=include_archived)
    return [TaskResponse.model_validate(task) for task in tasks]


@router.get("/{task_id}", response_model=TaskResponse)
async def get_task_route(
    task_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> TaskResponse:
    task = await get_task_or_404(db, current_user.id, task_id)
    return TaskResponse.model_validate(task)


@router.patch("/{task_id}", response_model=TaskResponse)
async def update_task_route(
    task_id: UUID,
    payload: TaskUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> TaskResponse:
    task = await get_task_or_404(db, current_user.id, task_id)
    updated = await update_task(db, task, payload)
    return TaskResponse.model_validate(updated)


@router.delete("/{task_id}", status_code=status.HTTP_204_NO_CONTENT)
async def archive_task_route(
    task_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> Response:
    task = await get_task_or_404(db, current_user.id, task_id)
    await archive_task(db, task)
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: frontend/lib/api/tasks.ts
import type { TaskCreate, TaskResponse, TaskUpdate } from "@/types/task"
import { apiClient } from "./client"

export async function getTasks(params?: {
  status?: string
  skip?: number
  limit?: number
}): Promise<TaskResponse[]> {
  const res = await apiClient.get<TaskResponse[]>("/api/v1/tasks", { params })
  return res.data
}

export async function getTask(id: string): Promise<TaskResponse> {
  const res = await apiClient.get<TaskResponse>(`/api/v1/tasks/${id}`)
  return res.data
}

export async function createTask(data: TaskCreate): Promise<TaskResponse> {
  const res = await apiClient.post<TaskResponse>("/api/v1/tasks", data)
  return res.data
}

export async function updateTask(id: string, data: TaskUpdate): Promise<TaskResponse> {
  const res = await apiClient.patch<TaskResponse>(`/api/v1/tasks/${id}`, data)
  return res.data
}

export async function deleteTask(id: string): Promise<void> {
  await apiClient.delete(`/api/v1/tasks/${id}`)
}
